README.md
# mcollective-zeromq-connector

This a connector for MCollective that communicates using the 0mq library.

You'll need a broker to connect things together, there's a simple ruby
implementation of this in [scripts](scripts)/mc0d.rb and a more perfomant
C++11 implementation available as [mc0d](https://github.com/puppetlabs/mc0d)

# Requirements

0MQ 4.0.x or later http://zeromq.org/intro:get-the-software

ffi-rzmq http://rubygems.org/gems/ffi-rzmq

MCollective 2.5.0 or later

# Configuration

The most basic configuration will look like this:

```
connector = zeromq
plugin.zeromq.broker.host = 192.168.1.1
plugin.zeromq.broker.public_key = /etc/mcollective/broker.public
plugin.zeromq.public_key = /etc/mcollective/this_actor.public
plugin.zeromq.private_key = /etc/mcollective/this_actor.private
```

The keys can be generated with [curve_keygen][curve_keygen] from the zeromq
distribution.  Save the two parts in .public and .private files, and distribute
the broker public key to all the nodes you wish to connect.

[curve_keygen]: https://github.com/zeromq/zeromq4-x/blob/master/tools/curve_keygen.c

# Settings

## plugin.zeromq.broker.host

String, required. The hostname or ip address of the host running the mc0d
broker.

## plugin.zeromq.broker.port

Number, optional (default is 61616).  The port to connect to the mc0d broker
on.

## plugin.zeromq.curve

Boolean, optional (default is true).  Should
[CURVE](http://rfc.zeromq.org/spec:25) encryption/authentication be enabled.
It is *strongly* suggested that you do not disable this.

## plugin.zeromq.broker.public_key

String, required (unless `plugin.zeromq.curve` is false).  Path to a file
containing the public key for the mc0d broker.

## plugin.zeromq.public_key

String, required (unless `plugin.zeromq.curve` is false). Path to a file
containing the public key for this client/server.

## plugin.zeromq.private_key

String, required (unless `plugin.zeromq.curve` is false).  Path to a file
containing the private key for this client/server.

## plugin.zeromq.heartbeat

Integer, optional (default: 30).  The interval, expressed in seconds, at which
to a) expect to hear from the broker, or b) send some traffic to the broker.
This is used to detect stalled/dead connections.

# Limitations and notes

For a description of the protocol in use please see [PROTOCOL.md](PROTOCOL.md).

zeromq's implementation of Curve does not appear to easily allow for trust
relationship between keys, so there's no existing model for 'create a ca to
issue keys and track revocations'.

zeromq additionally does not seem to allow for the validation of client keys,
by the server http://hintjens.com/blog:36, the clients can only verify the
server is the one indicated by `plugin.zeromq.broker.public_key`.

The implementation of the broker does not allow for persistent queues, and so
advanced uses of MCollective's asyncronous reply handling pattern such as
those explained [here][async_handling] are not currently supported.

[async_handling]: http://www.devco.net/archives/2012/08/19/mcollective-async-result-handling.php


Changes.md
# 0.0.1 Tuesday 13th May, 2014

* Initial public release


spec/spec_helper.rb
# fake up bits of MCollective
module MCollective
  class MessageNotReceived < Exception
  end
  class Log
  end
  module Connector
    class Base
    end
  end
end


spec/unit/mcollective/connector/zeromq_spec.rb
require 'spec_helper'
require 'mcollective/connector/zeromq'

module MCollective
  module Connector
    describe Zeromq do
      let(:pluginconf) do
        pluginconf = double('pluginconf')
        pluginconf.stub(:include?)
        pluginconf
      end

      let(:config) do
        conf = double('config')
        conf.stub(:identity).and_return("rspec-identity")
        conf.stub(:pluginconf).and_return(pluginconf)
        conf
      end

      let(:connector) do
        Zeromq.new
      end

      before :each do
        Config.stub(:instance).and_return(config)
        Zeromq.any_instance.stub(:get_option).with('zeromq.broker.host').and_return('rspec.host')
        Zeromq.any_instance.stub(:get_option).with('zeromq.broker.port', '61616').and_return('61617')
        Zeromq.any_instance.stub(:get_option).with('zeromq.heartbeat', '30').and_return('20')
        Zeromq.any_instance.stub(:get_bool_option).with('zeromq.curve', true).and_return(true)
        Zeromq.any_instance.stub(:get_option).with('zeromq.broker.public_key').and_return('middleware.public')
        Zeromq.any_instance.stub(:get_option).with('zeromq.public_key').and_return('instance.public')
        Zeromq.any_instance.stub(:get_option).with('zeromq.private_key').and_return('instance.private')
        IO.stub(:read).with('instance.public').and_return('instance public key contents')
        IO.stub(:read).with('instance.private').and_return('instance private key contents')
        IO.stub(:read).with('middleware.public').and_return('middleware public key contents')
        Log.stub(:debug)
      end

      describe '#initialize' do
        it 'should attempt to load the curve key' do
          connector.instance_variable_get(:@curve_enabled).should == true
          connector.instance_variable_get(:@broker_public).should == 'middleware public key contents'
          connector.instance_variable_get(:@our_public).should == 'instance public key contents'
          connector.instance_variable_get(:@our_private).should == 'instance private key contents'
        end

        it 'should not load curve when disabled' do
          Zeromq.any_instance.stub(:get_bool_option).with('zeromq.curve', true).and_return(false)
          connector.instance_variable_get(:@curve_enabled).should == false
          connector.instance_variable_get(:@broker_public).should == nil
          connector.instance_variable_get(:@our_public).should == nil
          connector.instance_variable_get(:@our_private).should == nil
        end

        it 'should zero out the @subscriptions' do
          connector.instance_variable_get(:@subscriptions).should == []
        end

        it 'should set @endpoint' do
          connector.instance_variable_get(:@endpoint).should == 'tcp://rspec.host:61617'
        end

        it 'should set @heartbeat' do
          connector.instance_variable_get(:@heartbeat).should == 20
        end
      end

      describe '#connect' do
        let(:socket) do
          socket = double('zeromq socket')
          socket.stub(:setsockopt).and_return(0)
          socket.stub(:connect).and_return(0)
          socket.stub(:send_strings).and_return(0)
          socket
        end

        before :each do
          ZMQ::Context.any_instance.stub(:socket).and_return(socket)
          Zeromq.any_instance.stub(:send_message)
          Zeromq.any_instance.stub(:recv_message).and_return([])
          Zeromq.any_instance.stub(:expect_ok_with)
          Zeromq.any_instance.stub(:start_heartbeat_threads)
        end

        it 'should create a new socket, configure and connect' do
          ZMQ::Context.any_instance.should_receive(:socket).with(ZMQ::DEALER).and_return(socket)
          socket.should_receive(:setsockopt).with(ZMQ::IDENTITY, "rspec-identity #{$$} #{Thread.current.inspect}")
          socket.should_receive(:setsockopt).with(ZMQ::LINGER, 0)
          socket.should_receive(:setsockopt).with(ZMQ::CURVE_SERVERKEY, 'middleware public key contents')
          socket.should_receive(:setsockopt).with(ZMQ::CURVE_PUBLICKEY, 'instance public key contents')
          socket.should_receive(:setsockopt).with(ZMQ::CURVE_SECRETKEY, 'instance private key contents')
          socket.should_receive(:connect).with('tcp://rspec.host:61617')
          connector.should_receive(:make_command_id).and_return('rspec-command-id')
          connector.should_receive(:send_message).with([ 'CONNECT' ] + {
            'VERSION' => '0.2',
            'TTL' => '20000',
            'ID' => 'rspec-command-id',
          }.to_a.sort.flatten)
          connector.should_receive(:expect_ok_with).with('rspec-command-id')
          connector.should_receive(:start_heartbeat_threads)

          connector.connect
        end

        it 'should not set up CURVE when disabled' do
          Zeromq.any_instance.stub(:get_bool_option).with('zeromq.curve', true).and_return(false)
          socket.should_not_receive(:setsockopt).with(ZMQ::CURVE_SERVERKEY, 'middleware public key contents')

          connector.connect
        end
      end
    end
  end
end


scripts/mc0d.rb
#!/usr/bin/env ruby
require 'ffi-rzmq'
require 'logger'
require 'optparse'

class Middleware
  class Client
    attr_accessor :id
    attr_accessor :ttl
    attr_accessor :subscriptions
    attr_accessor :last_recv, :next_recv_by
    attr_accessor :last_send, :next_send_by
  end

  def initialize(options)
    @recv_count = 0
    @send_count = 0
    @clients = {}

    $stdout.sync = true
    @logger = Logger.new(STDOUT)
    @logger.level = Logger::INFO

    @context = ZMQ::Context.new
    @socket = @context.socket(ZMQ::ROUTER)
    @socket_mutex = Mutex.new

    if options.include?(:curve_private_key)
      @logger.debug "enabling curve with private key '#{options[:curve_private_key]}'"
      private_key = IO.read(options[:curve_private_key]).chomp
      assert_zeromq(@socket.setsockopt(ZMQ::CURVE_SERVER, 1))
      assert_zeromq(@socket.setsockopt(ZMQ::CURVE_SECRETKEY, private_key))
    end

    # start listening to the network
    endpoint = "tcp://#{options[:bind]}:#{options[:port]}"
    @logger.debug "binding #{endpoint}"
    assert_zeromq(@socket.bind(endpoint))

    Thread.new do
      status_thread
    end
  end

  def main_loop
    @logger.info "going into mainloop"
    start_heartbeat_threads

    poller = ZMQ::Poller.new
    assert_zeromq(poller.register_readable(@socket))

    while true
      # TODO - figure out next ping we to get, and set polling based on this
      @logger.debug "polling for messages"
      poller.poll
      @logger.debug "poller woke up"

      if poller.readables.empty?
        # Here is whete we should put the sending of PING
        @logger.debug 'time to send some pings'
      else
        @logger.debug 'message available'
        # Message to route
        source, message = read_message

        client = @clients[source]
        if client.nil?
          client = Client.new
          client.id = source
          client.subscriptions = []
          @clients[client.id] = client
        end

        begin
          verb = message.shift
          if end_of_headers = message.index('')
            headers = Hash[*message.first(end_of_headers)]
            body = message.drop(end_of_headers + 1)
          else
            headers = Hash[*message]
            body = []
          end

          if headers['ID']
            id = [ 'ID', headers['ID'] ]
          else
            id = []
          end

          case verb
          when 'CONNECT'
            # TODO(richardc): versioncmp
            @logger.info "CONNECT '#{source}' #{headers.inspect}"
            if headers['TTL']
              client.ttl = Integer(headers['TTL'])
              if client.ttl
                client.next_send_by = Time.now + (0.95 * (client.ttl / 1000.0))
              end
            end

          when 'DISCONNECT'
            @logger.info "DISCONNECT '#{source}'"
            @clients.delete(client.id)

          when 'SUB'
            client.subscriptions += body

          when 'UNSUB'
            client.subscriptions -= body

          when 'PUT'
            topic = headers['TOPIC']
            count = 0
            payload = [ 'MESSAGE' ] + headers.reject{ |k,v| k == 'ID' }.to_a.flatten + [ '' ] + body
            @clients.values.select { |x| x.subscriptions.include?(topic) }.each do |subscriber|
              count += 1
              send_message(subscriber, payload)
            end
            @logger.info "forwarded message on '#{topic}' from '#{client.id}' to #{count} recipients"

          when 'NOOP'
            # yeah, whatever

          else
            @logger.error "Don't know how to handle '#{verb}', saying ERROR"
            send_message(client, [ 'ERROR', 'MESSAGE', 'Unknown verb' ] + id)
          end

          if !id.empty?
            # request for acknowledgement
            send_message(client, [ 'OK' ] + id)
          end

        rescue Exception => e
          @logger.error "Caught exception #{e}"
          send_message(client, [ 'ERROR', 'MESSAGE', "caught exception: #{e} #{e.backtrace}" ] + id)
        end

        client.last_recv = Time.now
        if client.ttl
          client.next_recv_by = Time.now + (1.05 * (client.ttl / 1000.0))
        end
      end
    end
  end

  private

  def status_thread
    @logger.info "Starting status thread"

    lastmsg = ''
    while true
      newmsg = "Currently managing #{@clients.count} clients. Recv: #{@recv_count}. Sent #{@send_count}."
      if lastmsg != newmsg
        @logger.info newmsg
        lastmsg = newmsg
      end
      sleep 2
    end
  end

  def read_message
    @logger.debug 'going to read message'
    message = []
    @socket_mutex.synchronize do
      assert_zeromq(@socket.recv_strings(message))
    end
    source = message.slice!(0,2)[0]

    @recv_count += 1
    @logger.debug "got #{message.inspect} from #{source.inspect}"
    return source, message
  end

  def send_message(client, message)
    @logger.debug "sending #{message.inspect} to #{client.id}"
    @socket_mutex.synchronize do
      @socket.send_strings([ client.id, '', *message ], ZMQ::DONTWAIT)
    end
    @logger.debug 'sent message'
    client.last_send = Time.now
    if client.ttl
      client.next_send_by = Time.now + (0.95 * (client.ttl / 1000.0))
    end
    @send_count += 1
  end

  def start_heartbeat_threads
    Thread.new { heartbeat_send_thread }
    Thread.new { heartbeat_recv_thread }
  end

  def heartbeat_send_thread
    begin
      while true
        @logger.debug 'heartbeat_send_thread: checking'
        now = Time.now
        @clients.values.select { |x| x.ttl && x.next_send_by }.each do |client|
          if now >= client.next_send_by
            @logger.info "heartbeat_send_thread: sending to #{client.id}"
            send_message(client, [ 'NOOP' ])
          end
        end
        first_to_send = @clients.values.collect { |x| x.ttl && x.next_send_by }.min

        sleep_for = first_to_send ? (first_to_send - now) : 1
        if sleep_for > 0
          @logger.debug "heartbeat_send_thread: sleeping for #{sleep_for} seconds"
          sleep sleep_for
        end
      end
    rescue Exception => e
      @logger.error "heartbeat_send_thread: #{e} #{e.backtrace}"
      retry
    end
  end

  def heartbeat_recv_thread
    begin
      while true
        now = Time.now
        @clients.values.select { |x| x.ttl && x.next_recv_by }.each do |client|
          if now >= client.next_recv_by
            @logger.info "heartbeat_recv_thread: '#{client.id}' has gone away, deleting"
            @clients.delete(client.id)
          end
        end
        first_to_recv = @clients.values.collect { |x| x.ttl && x.next_recv_by }.min

        sleep_for = first_to_recv ? (first_to_recv - now) : 1
        if sleep_for > 0
          @logger.debug "heartbeat_recv_thread: sleeping for #{sleep_for} seconds"
          sleep sleep_for
        end
      end
    rescue Exception => e
      @logger.error "heartbeat_send_thread: #{e}"
      retry
    end
  end

  def assert_zeromq(rc)
    return if ZMQ::Util.resultcode_ok?(rc)
    raise "Operation failed, errno [#{ZMQ::Util.errno}] description [#{ZMQ::Util.error_string}]"
  end
end


options = {
  :bind => '*',
  :port => 61616,
}
OptionParser.new do |opts|
  opts.banner = "Usage: middleware [options]"

  opts.on('--bind', 'Address to bind') do |v|
    options[:bind] = v
  end

  opts.on('--port', 'Port to bind') do |v|
    options[:port] = v
  end

  opts.on('--curve-private-key=s', 'The curve privatekey') do |v|
    options[:curve_private_key] = v
  end
end.parse!

Middleware.new(options).main_loop


lib/mcollective/connector/zeromq.rb
require 'ffi-rzmq'
require 'securerandom'

module MCollective
  module Connector
    class Zeromq < Base
      # Implementation notes:

      # This connector is a simple 0MQ connector that implements the 0.1
      # protocol as outlined in PROTOCOL.md

      # Topics used:
      #   "#{collective} #{agent}"                - broadcast for agents
      #   "#{collective} reply #{identity} #{$$}" - replies
      #   "#{collective} nodes #{identity}"       - direct addressing to node

      def initialize
        host = get_option('zeromq.broker.host')
        port = get_option('zeromq.broker.port', '61616')
        @endpoint = "tcp://#{host}:#{port}"
        @heartbeat = Integer(get_option('zeromq.heartbeat', '30'))

        @context = ZMQ::Context.new
        @socket = nil
        @socket_mutex = Mutex.new
        @subscriptions = []

        @curve_enabled =  get_bool_option('zeromq.curve', true)
        if @curve_enabled
          # load the curve keys
          @broker_public = IO.read(get_option('zeromq.broker.public_key')).chomp
          @our_public  = IO.read(get_option('zeromq.public_key')).chomp
          @our_private = IO.read(get_option('zeromq.private_key')).chomp
        end
      end

      def connect
        Log.debug("creating new socket")
        @socket = @context.socket(ZMQ::DEALER)

        # set an identity based on the identity, pid, and thread.id of this context
        assert_zeromq(@socket.setsockopt(ZMQ::IDENTITY, "#{Config.instance.identity} #{$$} #{Thread.current.inspect}"))

        # don't wait for messages to get delivered/received on close
        assert_zeromq(@socket.setsockopt(ZMQ::LINGER, 0))

        if @curve_enabled
          # key the socket
          @socket.setsockopt(ZMQ::CURVE_SERVERKEY, @broker_public)
          @socket.setsockopt(ZMQ::CURVE_PUBLICKEY, @our_public)
          @socket.setsockopt(ZMQ::CURVE_SECRETKEY, @our_private)
        end

        Log.debug("connecting @socket to #{@endpoint}'")
        assert_zeromq(@socket.connect(@endpoint))

        id = make_command_id
        options = {
          'VERSION' => '0.2',
          'TTL'     => (@heartbeat * 1000).to_s,
          'ID'      => id,
        }

        send_message([ 'CONNECT' ] + options.to_a.sort.flatten)
        expect_ok_with(id)

        if !@subscriptions.empty?
          # reconnection, resub
          id = make_command_id
          send_message([ 'SUB', 'ID', id, '' ] + @subscriptions)
          expect_ok_with(id)
        end

        start_heartbeat_threads
      end

      def disconnect(stop_threads = true)
        Log.debug('disconnecting')
        if stop_threads
          stop_heartbeat_threads
        end

        send_message([ 'DISCONNECT' ])
        sleep 0.5 # allow for DISCONNECT to be queued/delivered
        @socket_mutex.synchronize do
          assert_zeromq(@socket.close)
        end
        @socket = nil
        Log.debug('disconnected')
      end

      def subscribe(agent, type, collective)
        topic = topic_for_kind(agent, type, collective)

        Log.debug("subscribing to topic '#{topic}'")
        id = make_command_id
        send_message([ 'SUB', 'ID', id, '', topic ])
        expect_ok_with(id)
        @subscriptions += [ topic ]
      end

      def unsubscribe(agent, type, collective)
        topic = topic_for_kind(agent, type, collective)

        Log.debug("unsubscribing from topic '#{topic}'")
        id = make_command_id
        send_message([ 'UNSUB', 'ID', id, '', topic ])
        expect_ok_with(id)
        @subscriptions -= [ topic ]
      end

      def publish(message)
        Log.debug('publish')

        if message.type == :direct_request
          message.discovered_hosts.each do |node|
            publish_message(message, node)
          end
        else
          publish_message(message)
        end

        Log.debug("publish done")
      end

      def receive
        while true
          Log.debug('receive a message from zeromq')
          message = recv_message
          @last_recv = Time.now
          @next_recv_by = @last_recv + @heartbeat

          verb = message.shift
          case verb
          when 'NOOP'
            # silently accept

          when 'MESSAGE'
            if end_of_headers = message.index('')
              headers = Hash[*message.first(end_of_headers)]
              body = message.drop(end_of_headers + 1)
            else
              headers = Hash[*message]
              body = []
            end

            Log.debug("MESSAGE received with #{headers.inspect}")
            return Message.new(body[0], nil, :headers => headers)

          else
            Log.error("Unexpected message '#{verb}' from middleware: #{message.inspect}")
          end
        end
      end

      private

      # headers_for_message, headers_for_kind, and topic_for_kind ape the
      # behaviour of the activemq connector a little.  As it's fresh here's a
      # braindump
      #
      # The problem is you need to be able to compute a topic/queue name in a
      # certain number of situations:
      #    As a client
      #       when subscribing to your reply topic
      #       when sending a message to the servers to say 'please do this'
      #       when sending a directed message to a specific server to say 'please do this'

      #    As a daemon
      #       when subscribing to all the topics for all the agents you have
      #       when unsubscribing from those same set of topics
      #       when sending a response to a message

      # Subscription happens in response to a call to connector.subscribe(agent, type, collective)
      # and so these directly call topic_for_kind(agent, type, collective) and use the topic part of the answer

      # Unsubscription is the same pattern, but from connector.unsubscribe

      # Sending any request via .publish(MCollective::Message) is where it
      # gets interesting.

      # In the 'broadcast' mode we call topic_for_message(message) which then
      # (assuming it's not a reply), delegates to topic_for_kind(message.agent, message.type, message.collective)

      # If it's a directed request from direct addressing we iterate over the
      # set of discovered nodes (message.discovered_hosts) and call
      # topic_for_message(message, target) which calls topic_for_kind(a,t,c,
      # NODE) to generate a message with (in the activemq case) a header that
      # will match the selector.  With this connector we just make another
      # topic per node, with the assumption that it's possibly cheap enough
      # for ZeroMQ to have 2,000 'topics'

      # In all the 'we're making a request' codepaths (:broadcast, :request)
      # we also return a set of headers with our reply_to set.  This is the
      # mechanism by which we can send a response at all, and is in no other
      # part of the message. (it's message.request.headers['reply_to'])

      # the headers we should use when sending this message
      def headers_for_message(message, target = nil)
        case message.type
        when :reply
          {
            'TOPIC' => message.request.headers['X-REPLY-TO'],
          }
        else
          headers_for_kind(message.agent, message.type, message.collective, target)
        end
      end

      def headers_for_kind(agent, type, collective, target = nil)
        reply_to = "#{collective} reply #{Config.instance.identity} #{$$}"
        headers = {}
        case type
        when :reply
          headers['TOPIC'] = reply_to

        when :broadcast, :request
          headers['TOPIC'] = "#{collective} #{agent}"
          headers['X-REPLY-TO'] = reply_to

        when :direct_request
          # When we send a directed message
          headers['TOPIC'] = "#{collective} nodes #{target}"
          headers['X-REPLY-TO'] = reply_to

        when :directed
          # Where we listen for directed messages
          headers['TOPIC'] = "#{collective} nodes #{Config.instance.identity}"

        else
          raise "Unknown message type #{type}"
        end

        headers
      end

      def topic_for_kind(agent, type, collective)
        headers_for_kind(agent, type, collective)['TOPIC']
      end

      def publish_message(message, node = nil)
        headers = headers_for_message(message, node)

        if node
          Log.debug("sending direct message to '#{node}' with headers '#{headers.inspect}'")
        else
          Log.debug("sending broadcast message with headers '#{headers.inspect}'")
        end

        send_message([ 'PUT' ] + headers.to_a.flatten + [ '', message.payload ])
      end

      def send_message(message)
        Log.debug "sending #{message.inspect}"
        @socket_mutex.synchronize do
          assert_zeromq(@socket.send_strings([''] + message, ZMQ::DONTWAIT))
        end
        @next_send_by = Time.now + (@heartbeat * 0.95)
        Log.debug 'sent'
      end

      def recv_message
        response = []
        poller = ZMQ::Poller.new
        poller.register_readable(@socket)
        poller.poll
        if poller.readables.empty?
          # We probably woke up because the connection was closed.  Raise a backoff error
          raise MessageNotReceived.new(10), "Socket didn't become readable"
        end

        @socket_mutex.synchronize do
          Log.debug 'doing read'
          assert_zeromq(@socket.recv_strings(response))
        end
        @last_recv = Time.now
        @next_recv_by = Time.now + (@heartbeat * 1.05)
        # as we're now a DEALER first frame is ''
        response.shift
        Log.debug "got #{response.inspect}"
        return response
      end

      def start_heartbeat_threads
        if !@heartbeat_send_thread
          Log.debug("spawning the heartbeat_send_thread")
          @heartbeat_send_thread = Thread.new { heartbeat_send_thread }
        end

        @last_recv = nil
        @next_recv_by = Time.now + @heartbeat
        if !@heartbeat_recv_thread
          Log.debug("spawning the heartbeat_expected_thread")
          @heartbeat_recv_thread = Thread.new { heartbeat_recv_thread }
        end
      end

      def stop_heartbeat_threads
        if @heartbeat_send_thread
          @heartbeat_send_thread.kill
          @heartbeat_send_thread = nil
        end

        if @heartbeat_recv_thread
          @heartbeat_recv_thread.kill
          @heartbeat_recv_thread = nil
        end
      end

      # This thread ensure we keep saying something
      def heartbeat_send_thread
        begin
          while true
            if Time.now >= @next_send_by
              Log.debug "heartbeat_send_thread: announcing ourself to the world"
              send_message([ 'NOOP' ])
            end

            sleep_for = @next_send_by - Time.now
            if sleep_for > 0
              Log.debug "heartbeat_send_thread: sleeping #{sleep_for} seconds"
              sleep sleep_for
            end
          end
        rescue Exception => e
          Log.error "heartbeat_send_thread: #{e}"
          retry
        end
      end

      # This thread expects messages to come to us, reconnects if it didn't
      def heartbeat_recv_thread
        begin
          while true
            if Time.now >= @next_recv_by
              last_heard = 'never'
              if @last_recv
                last_heard = "#{Time.now - @last_recv} seconds ago"
              end

              Log.warn "Last heard from the middleware #{last_heard}.  Reconnecting."
              # disconnect(false), don't shoot the keepalive thread (ourself) in the face
              disconnect(false)
              sleep 1
              connect
            end

            sleep_for = @next_recv_by - Time.now
            if sleep_for > 0
              Log.debug "heartbeat_recv_thread: sleeping for #{sleep_for} seconds"
              sleep sleep_for
            end
          end
        rescue Exception => e
          Log.error "heartbeat_send_thread: #{e}"
          retry
        end
      end

      def make_command_id
        SecureRandom.uuid
      end

      def expect_ok_with(id)
        message = recv_message
        verb = message.shift
        headers = Hash[*message]
        if verb != 'OK' || headers['ID'] != id
          raise MessageNotReceived.new(10), "Didn't receive the expected OK frame"
        end
      end

      def assert_zeromq(rc)
        return if ZMQ::Util.resultcode_ok?(rc)
        raise "zeromq operation failed, errno [#{ZMQ::Util.errno}] description [#{ZMQ::Util.error_string}]"
      end

      def get_option(opt, default=nil)
        if Config.instance.pluginconf.include?(opt)
          return Config.instance.pluginconf[opt]
        end

        return default unless default.nil?

        raise("No plugin.#{opt} configuration option given")
      end

      def get_bool_option(opt, default=nil)
        if Config.instance.pluginconf.include?(opt)
          return Util.str_to_bool(Config.instance.pluginconf[opt])
        end

        return default unless default.nil?

        raise("No plugin.#{opt} configuration option given")
      end
    end
  end
end
